flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp

```diff
@@ -712,17 +757,66 @@ static mlir::Value generateReductionLoop(fir::FirOpBuilder &builder,
 ///     %c = <reduce count> %i
 ///     fir.result %c
 template <typename Op>
-class ReductionElementalConversion : public mlir::OpRewritePattern<Op> {
+class ReductionConversion : public mlir::OpRewritePattern<Op> {
 public:
   using mlir::OpRewritePattern<Op>::OpRewritePattern;
 
   llvm::LogicalResult
   matchAndRewrite(Op op, mlir::PatternRewriter &rewriter) const override {
     mlir::Location loc = op.getLoc();
-    hlfir::ElementalOp elemental =
-        op.getMask().template getDefiningOp<hlfir::ElementalOp>();
-    if (!elemental || op.getDim())
-      return rewriter.notifyMatchFailure(op, "Did not find valid elemental");
+    // Select source and validate its arguments.
+    mlir::Value source;
+    bool valid = false;
+    if constexpr (std::is_same_v<Op, hlfir::AnyOp> ||
+                  std::is_same_v<Op, hlfir::AllOp> ||
+                  std::is_same_v<Op, hlfir::CountOp>) {
+      source = op.getMask();
+      valid = !op.getDim();
+    } else if constexpr (std::is_same_v<Op, hlfir::MaxvalOp> ||
```

Comment:
The mask operand is not handled, so we should disallow the optimization when it is present.

Reply:
I believe we should keep this PR minimum. Supporting all the cases will follow later on.

>The mask operand is not handled

It is complicated, but any/all/count's default argument is named mask. This operand acts as a mask over itself.

Reply:
I meant that the `maxval` and `minval` operations may have a mask operand.  I do not see this operand being handled anywhere (maybe I missed it).  We cannot just ignore it, so I suggest that we disallow the optimization if this operand is present.

Reply:
Thank you for your review. I was checking a wrong line. `!op.getDim() && !op.getDim();` was meant to be `!op.getDim() && !op.getMask()`.